### DISCUSSION TITLE
Creating nonlinear variables in global user object

### DISCUSSION THREAD

**Discussion by rtaylo45** [idaholab/moose#18483] [Q&A General] [answered]
Hello,

I created a user object that hold information about specific species such as molar mass, species name, phase... etc. The sample code in the input block is:

```
[UserObjects]
  [spec_def]
    type = SpeciesDefinition
    species_names = 'U-235 Xe-135 Xe-135 I-135'
    phase_names = 'liq gas gas liq'
    molar_masses  = '235 135 135 135'
  []
[]
```
and i would like for the user block to automatically create the nonlinear variables and add them to the problem but i can not figure out a way to go about this. This is because the `GeneralUserObject` does not have access to the _factory object or _problem object. I have tried inheriting my object from the Action class as well. But this doesn't really work, lots of compiler errors with multiple functions being defined with `GeneralUserObject` and `Action`. I have tried looking at examples in the `Geochemistry` and `chemical reactions` modules but they make the user add in the variables themselves. The problem is that in the future i will be needing to add hundreds of species and i don't want the input file to be super long, might switch to an addition input file for that. So my question is: Can we create nonlinear variables and add them to the problem using GeneralUserObjects? 

If this is not the case, the other idea i had was to create a `AddSpeciesAction` that takes in my `SpeciesDefinition` user object as a parameter. It would then iterate though the species and add them as variables. But im not sure how to add parameters of custom types. Is there a way to add parameters of custom types?

Thanks,
Zack

**Comment by GiudGiud** [comment, marked as answer]
Hello

The cannon way to add non linear variables to the problem like this is to define a new action. You can keep the same syntax in the input file, just replace [UserObjects] by a new category with an associated action that does the non linear variable creation. 

I m not sure about adding a UserObject parameter to this action with a custom type. I would pass a string (the name) then go retrieve the UO from the action.

Guillaume

**Reply by dschwen** [reply]
To expand on @GiudGiud's comment: Create both the variables _and_ the user object in the action. Only actions can (should) create other objects, and you cannot have an object derive from multiple systems.
